### DISCUSSION TITLE
EAllowedDOFs and constraints

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#685] [Q&A] [answered]
I'm guessing you're already aware of this, so I figured I'd make this a discussion rather than an issue, but when using `EAllowedDOFs` to lock various degrees-of-freedom while also attaching such a body to a various constraints, you can end up triggering asserts like this one:

https://github.com/jrouwe/JoltPhysics/blob/5d382fc0f71411dacf0c1f6f93fe0d2612fed45e/Jolt/Physics/Constraints/ConstraintPart/HingeRotationConstraintPart.h#L112-L116

... ~~and end up with NaNs all over the place~~.

This exact problem happens with Godot Physics as well, when using the [`lock_rotation`](https://docs.godotengine.org/en/stable/classes/class_rigidbody3d.html#class-rigidbody3d-property-lock-rotation) property, as that similarly zeroes out the inverse inertia. However, when instead locking individual axes using the [`axis_lock_*`](https://docs.godotengine.org/en/stable/classes/class_physicsbody3d.html#class-physicsbody3d-property-axis-lock-angular-x) properties, Godot Physics instead uses the less-than-ideal method of just zeroing out the velocities, avoiding the issue with constraints, whereas I rely on `EAllowedDOFs` for this, resulting in a somewhat unfortunate discrepancy.

Do you have any ideas for mitigations here? I'm having a hard time imagining that I could somehow temporarily allow for certain DOFs while connected to certain constraints, so I'm leaning more towards just having all bodies connected to constraints be forced to use `EAllowedDOFs::All`, and emitting a warning accordingly.

EDIT: My bad, it seems I mistook the assertion in `HingeRotationConstraintPart` to be a precursor for the NaN issue. The NaN issues seems to be specific to `SixDOFConstraint` and potentially a couple of others, where it instead triggers the `!mAngularVelocity.IsNaN()` assertion here:

https://github.com/jrouwe/JoltPhysics/blob/5d382fc0f71411dacf0c1f6f93fe0d2612fed45e/Jolt/Physics/Constraints/ConstraintPart/RotationEulerConstraintPart.h#L54

Maybe `RotationEulerConstraintPart::CalculateConstraintProperties` needs to call `Deactivate`, similar to `HingeRotationConstraintPart`, if possible?

**Comment by jrouwe** [comment, marked as answer]
Can you check if #686 fixes the issue?

**Reply by mihe** [reply]
That does seem to fix the issue, for the cases I've run into.

**Reply by jrouwe** [reply]
Cool, thanks for testing! Change has been merged into master.

**Reply by mihe** [reply]
Fantastic, thank you!
